### PR TITLE
Dockerization

### PR DESCRIPTION
**Pull request by YuYanzy**
- [x] dockerized current development environment
- [X] added documentation about setup process

### CODE REVIEW

**Review by JackHKJ — APPROVED** [review]
All Changes Reviewed

File: src/backend/.env
Comment (JackHKJ) [inline src/backend/.env:1]: Should this file be committed or added to gitignore and config manually?

Comment (YuYanzy) [reply src/backend/.env:1]: I think it should be tracked by git as environment variables are also worth tracking.

File: .gitignore [hunk 1]
```diff
@@ -107,7 +107,7 @@ celerybeat.pid
 
 # Environments
 ENV.py
-.env
+# .env
```
Comment (Windsooon) [inline .gitignore:110]: We can delete line 109 and line 110 in the PR

File: src/backend/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#base image
```
Comment (Windsooon) [inline src/backend/Dockerfile:1]: How about

    # Set up base image for docker

File: src/backend/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+FROM python:3.11
+# set work directory
+WORKDIR /app
+
+# installation of dependencies
```
Comment (Windsooon) [inline src/backend/Dockerfile:6]: Try

    # Copy and install dependencies

File: src/backend/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+COPY requirements.txt /app/requirements.txt
+RUN pip install -r requirements.txt
+
+# copy project
```
Comment (Windsooon) [inline src/backend/Dockerfile:10]: Comments in Line 10 and Line 12 are not necessary here. The code is self-explained

File: src/backend/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# How to Set Up
+
+## Set Up with Docker 
+If you already have Docker and docker-compose installed on your local environment, you can directly use the following command to build and start the project. (Make sure your Docker Daemon is running)
+```Shell
+chmod +x run.sh
```
Comment (Windsooon) [inline src/backend/README.md:6]: What are the benefits of using shell commands over executing docker-compose directly?

File: src/backend/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+```
+
+After that, you can modify your code normally and all changes will be updated to the Docker container.
+
+## Customize Docker configuration and environment variables
+Please follow comments in `docker-compose.yml` and `Dockerfile` to configure your Docker configuration.
+For environment variables, you can modify them in `.env` file.
+
+## How to stop
+Using `Cltr-C` or `CMD-C` to kill the `run.sh` process can stop the docker container and project.
```
Comment (Windsooon) [inline src/backend/README.md:24]: Same question, What are the benefits of using shell commands over executing docker-compose-related commands directly?

File: src/backend/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+services:
+  backend:
+    container_name: educationx_backend
+    build:    # build image
+      context: .
+      dockerfile: Dockerfile
+    working_dir: /app  # working directory inside container
+    env_file:     # environment variables
+      - .env
+    volumes:       # enable hot reload
```
Comment (Windsooon) [inline src/backend/docker-compose.yml:12]: What do you mean `# enable hot reload` when using volumes?

File: src/backend/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      - .:/app
+      - ../static:/static
+    ports:         # port mapping
+      - "8000:8000"
+    command: [ sh, -c, 'sleep 10 &&  python manage.py migrate &&  python manage.py runserver 0.0.0.0:8000'] # first time run
+    # command: [ sh, -c, 'sleep 10 &&  python manage.py runserver 0.0.0.0:8000'] # regular run
```
Comment (Windsooon) [inline src/backend/docker-compose.yml:18]: we can remove the useless code here.

File: src/backend/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      dockerfile: Dockerfile
+    working_dir: /app  # working directory inside container
+    env_file:     # environment variables
+      - .env
+    volumes:       # enable hot reload
+      - .:/app
+      - ../static:/static
+    ports:         # port mapping
+      - "8000:8000"
+    command: [ sh, -c, 'sleep 10 &&  python manage.py migrate &&  python manage.py runserver 0.0.0.0:8000'] # first time run
```
Comment (Windsooon) [inline src/backend/docker-compose.yml:17]: I'm not sure `sleep 10` is a good practice here. Maybe you can give me some resources so I can check it out later.

**Review by Windsooon — CHANGES_REQUESTED** [review]

File: src/backend/Dockerfile (hunk 4 above)
Comment (Mive667) [reply src/backend/Dockerfile:10]: how about keep them for now, and we could delete all of dead code and unnecessary comments when the project is almost down?

**Pull request closed without merging**